tests/transforms/test_decompose_transform.py

```diff
@@ -0,0 +1,240 @@
+                qml.T(wires=[0]),
+                qml.ops.op_math.Adjoint(qml.T(wires=[1])),
+                qml.CNOT(wires=[0, 1]),
+            ],
+            None,
+        ),
+    ]
+
+    @pytest.mark.parametrize("gate_set", gate_set_inputs)
+    def test_different_input_formats(self, gate_set):
+        """Tests that gate sets of different types are handled correctly"""
+        tape = qml.tape.QuantumScript([qml.RX(0, wires=[0])])
+        decompose(tape, gate_set=gate_set)
+
+    def test_user_warning(self):
+        """Tests that user warning is raised if operator does not have a valid decomposition"""
+        tape = qml.tape.QuantumScript([qml.RX(0, wires=[0])])
+        with pytest.warns(UserWarning, match="has no supported decomposition"):
+            decompose(tape, gate_set=lambda op: op.name not in {"RX"})
+
+    def test_infinite_decomposition_loop(self):
+        """Test that a recursion error is raised if decomposition enters an infinite loop."""
+        tape = qml.tape.QuantumScript([InfiniteOp(1.23, 0)])
+        with pytest.raises(RecursionError, match=r"Reached recursion limit trying to decompose"):
+            decompose(tape, lambda obj: obj.has_matrix)
+
+    @pytest.mark.parametrize("initial_ops, gate_set, expected_ops, warning_pattern", iterables_test)
+    def test_iterable_gate_set(self, initial_ops, gate_set, expected_ops, warning_pattern):
+        """Tests that gate sets defined with iterables decompose correctly"""
+        tape = qml.tape.QuantumScript(initial_ops)
+
+        if warning_pattern is not None:
+            with pytest.warns(UserWarning, match=warning_pattern):
+                (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
+        else:
+            (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
+
+        expected_tape = qml.tape.QuantumScript(expected_ops)
+
+        qml.assert_equal(decomposed_tape, expected_tape)
+
+    @pytest.mark.parametrize("initial_ops, gate_set, expected_ops, warning_pattern", iterables_test)
+    def test_callable_gate_set(self, initial_ops, gate_set, expected_ops, warning_pattern):
+        """Tests that gate sets defined by callables decompose correctly"""
+        tape = qml.tape.QuantumScript(initial_ops)
+
+        if warning_pattern is not None:
+            with pytest.warns(UserWarning, match=warning_pattern):
+                (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
+        else:
+            (decomposed_tape,), _ = decompose(tape, gate_set=gate_set)
+
+        expected_tape = qml.tape.QuantumScript(expected_ops)
+
+        qml.assert_equal(decomposed_tape, expected_tape)
+
+
+def test_null_postprocessing():
+    """Tests the null postprocessing function in the decompose transform"""
+    tape = qml.tape.QuantumScript([qml.Hadamard(0), qml.RX(0, 0)])
+    (new_tape,), fn = qml.transforms.decompose(tape, gate_set={qml.RX, qml.RZ})
+    qml.assert_equal(fn(new_tape), new_tape.circuit[0])
```

Comment:
```suggestion
    assert fn((1,  ) ) == 1
```

This kinda makes it look like the postprocessing needs to act on the circuit. While technically valid for checking that the postprocessing works, it might be clearer to read to check how it works on something that looks like a result.

So the `fn` is for any postprocessing after we've calculated results.  This gets used in things like `split_non_commuting` and `broadcast_expand`, where we have to restructure the result object.